# a.

In [1]:
!pip install mlxtend

  Using cached mlxtend-0.20.0-py2.py3-none-any.whl (1.3 MB)


In [26]:
import pandas as pd
from mlxtend.preprocessing import TransactionEncoder
from mlxtend.frequent_patterns import apriori, association_rules


# b.

In [2]:
df = pd.read_csv('datasets/Hackathon_Working_Data.csv')

In [3]:
df.shape

(26985, 14)

In [4]:
df.head()

MONTH STORECODE  DAY BILL_ID  BILL_AMT  QTY  VALUE  PRICE  \
0    M1        N1    4    T375     225.0  1.0  225.0  225.0   
1    M1        N1    4    T379      95.0  1.0   95.0   95.0   
2    M1        N1    4    T381      10.0  1.0   10.0   10.0   
3    M1        N1    4    T382     108.0  1.0  108.0  108.0   
4    M1        N1    4    T384      19.0  1.0   19.0   19.0   

                       GRP                     SGRP                    SSGRP  \
0     BUTTER MARGR  (4/94)                   BUTTER                   SALTED   
1  CONFECTIONERY - ECLAIRS  CONFECTIONERY - ECLAIRS  CONFECTIONERY - ECLAIRS   
2                CHOCOLATE         CHOCOLATE PANNED         CHOCOLATE PANNED   
3             PACKAGED TEA               MAIN PACKS               MAIN PACKS   
4         ALL IODISED SALT            POWDERED SALT            POWDERED SALT   

                      CMP           MBRD                 BRD  
0               G C M M F           AMUL                AMUL  
1             PARLE PRODS         MELODY    MELODY CHOCOLATY  
2  MONDELEZ INTERNATIONAL  CADBURY SHOTS       CADBURY SHOTS  
3      GUJ TEA PROCESSORS     WAGH BAKRI  WAGH BAKRI INSTANT  
4               TATA CHEM           TATA           TATA SALT

In [5]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 26985 entries, 0 to 26984
Data columns (total 14 columns):
 #   Column     Non-Null Count  Dtype  
---  ------     --------------  -----  
 0   MONTH      26985 non-null  object 
 1   STORECODE  26985 non-null  object 
 2   DAY        26985 non-null  int64  
 3   BILL_ID    26985 non-null  object 
 4   BILL_AMT   26985 non-null  float64
 5   QTY        26985 non-null  float64
 6   VALUE      26985 non-null  float64
 7   PRICE      26985 non-null  float64
 8   GRP        26985 non-null  object 
 9   SGRP       26985 non-null  object 
 10  SSGRP      26985 non-null  object 
 11  CMP        26985 non-null  object 
 12  MBRD       26985 non-null  object 
 13  BRD        26985 non-null  object 
dtypes: float64(4), int64(1), object(9)
memory usage: 2.9+ MB


In [6]:
df.isna().sum()

MONTH        0
STORECODE    0
DAY          0
BILL_ID      0
BILL_AMT     0
QTY          0
VALUE        0
PRICE        0
GRP          0
SGRP         0
SSGRP        0
CMP          0
MBRD         0
BRD          0
dtype: int64

# c.

In [7]:
df['STORE_BILL_ID'] = df['STORECODE'] + '_' + df['BILL_ID']

In [8]:
df['STORE_BILL_ID']

0          N1_T375
1          N1_T379
2          N1_T381
3          N1_T382
4          N1_T384
           ...    
26980    N10_T5999
26981    N10_T6007
26982    N10_T6007
26983    N10_T6007
26984    N10_T6013
Name: STORE_BILL_ID, Length: 26985, dtype: object

# d.

In [9]:
sum_df = df.groupby(['STORE_BILL_ID','GRP'])['QTY'].agg({'sum'})

In [10]:
sum_df

sum
STORE_BILL_ID GRP                               
N10_T1038     SPICES (03/04)                 1.0
N10_T1039     BUTTER MARGR  (4/94)           1.0
N10_T1040     VERMICELLI & NOODLE            1.0
N10_T1043     CHOCOLATE                      1.0
N10_T1048     SPICES (03/04)                 2.0
...                                          ...
N9_T953       TOILET SOAPS (04/00)           1.0
              VANASPATI CONS PACK (8/94)     1.0
              WASHING POWDERS/LIQUIDS        4.0
N9_T96        CHEESE        (4/94)           2.0
N9_T99        REFINED EDIBLE OILS-ALL PACKS  1.0

[22379 rows x 1 columns]

# e.

In [30]:
df.QTY[df['QTY']>=1] = 1
df.QTY[df['QTY']<1] = 0

C:\Users\ACER\AppData\Local\Temp\ipykernel_20804\4260823984.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df.QTY[df['QTY']>=1] = 1
C:\Users\ACER\AppData\Local\Temp\ipykernel_20804\4260823984.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df.QTY[df['QTY']<1] = 0


# f.

In [13]:
df.groupby(["STORE_BILL_ID"])["GRP"].apply(list)

STORE_BILL_ID
N10_T1038                                     [SPICES (03/04)]
N10_T1039                               [BUTTER MARGR  (4/94)]
N10_T1040                                [VERMICELLI & NOODLE]
N10_T1043                                          [CHOCOLATE]
N10_T1048                     [SPICES (03/04), SPICES (03/04)]
                                   ...                        
N9_T940                                 [PACKAGED ATTA (1/95)]
N9_T95                                       [KETCHUPS/SAUCES]
N9_T953      [BISCUITS - CORE & NON CORE, BISCUITS - CORE &...
N9_T96                                  [CHEESE        (4/94)]
N9_T99                         [REFINED EDIBLE OILS-ALL PACKS]
Name: GRP, Length: 13491, dtype: object

In [14]:
te = TransactionEncoder()
te_ary = te.fit_transform(df.groupby(["STORE_BILL_ID"])["GRP"].apply(list))

In [17]:
new_df = pd.DataFrame(te_ary, columns=te.columns_)

In [22]:
new_df

AFTER SHAVE LOTIONS  AGARBATTI & DHOOPBATTI  ALL AIR FRESHNERS(01/03)  \
0                    False                   False                     False   
1                    False                   False                     False   
2                    False                   False                     False   
3                    False                   False                     False   
4                    False                   False                     False   
...                    ...                     ...                       ...   
13486                False                   False                     False   
13487                False                   False                     False   
13488                False                   False                     False   
13489                False                   False                     False   
13490                False                   False                     False   

       ALL IODISED SALT  ANTACIDS  ANTISEPTIC CREAMS  \
0                 False     False              False   
1                 False     False              False   
2                 False     False              False   
3                 False     False              False   
4                 False     False              False   
...                 ...       ...                ...   
13486             False     False              False   
13487             False     False              False   
13488             False     False              False   
13489             False     False              False   
13490             False     False              False   

       ANTISEPTIC LIQUIDS (4/97)  BABY FOODS  BABY OILS / MASSAGE OILS  \
0                          False       False                     False   
1                          False       False                     False   
2                          False       False                     False   
3                          False       False                     False   
4                          False       False                     False   
...                          ...         ...                       ...   
13486                      False       False                     False   
13487                      False       False                     False   
13488                      False       False                     False   
13489                      False       False                     False   
13490                      False       False                     False   

       BATTERIES  ...  SUGAR SUBSTITUTE (11/05)  TALCUM POWDERS  \
0          False  ...                     False           False   
1          False  ...                     False           False   
2          False  ...                     False           False   
3          False  ...                     False           False   
4          False  ...                     False           False   
...          ...  ...                       ...             ...   
13486      False  ...                     False           False   
13487      False  ...                     False           False   
13488      False  ...                     False           False   
13489      False  ...                     False           False   
13490      False  ...                     False           False   

       TOILET SOAPS (04/00)  TOOTH BRUSH   (7/93)  TOOTH PASTES  \
0                     False                 False         False   
1                     False                 False         False   
2                     False                 False         False   
3                     False                 False         False   
4                     False                 False         False   
...                     ...                   ...           ...   
13486                 False                 False         False   
13487                 False                 False         False   
13488                  True                 False         False   
13489    

In [23]:
frequent_itemsets2 = apriori(new_df, min_support=0.01, use_colnames=True)

In [24]:
frequent_itemsets2

support                                           itemsets
0   0.033356                                 (ALL IODISED SALT)
1   0.013861                                        (BEVERAGES)
2   0.283596                       (BISCUITS - CORE & NON CORE)
3   0.023720                                (BREAKFAST CEREALS)
4   0.038841                             (BUTTER MARGR  (4/94))
5   0.018457                             (CHEESE        (4/94))
6   0.090357                                        (CHOCOLATE)
7   0.020606                                 (CLEANERS - FLOOR)
8   0.012230                                (CLEANERS - TOILET)
9   0.046624                               (CLEANERS - UTENSIL)
10  0.020829                        (COCONUT OIL PKTYP (08/02))
11  0.011860                                           (COFFEE)
12  0.020236                       (CONFECTIONERY - TOFFEE/HBC)
13  0.024238                             (DETERGENT CAKES/BARS)
14  0.011267                             (DIAPERS       (8/93))
15  0.014380                             (HAIR DYES     (1/76))
16  0.014676                             (HAIR OILS PKTP(8/02))
17  0.024313                                  (KETCHUPS/SAUCES)
18  0.011044                      (LIQUID TOILET SOAPS  (8/93))
19  0.043214                             (PACKAGED ATTA (1/95))
20  0.023720                               (PACKAGED PURE GHEE)
21  0.069602                                     (PACKAGED TEA)
22  0.016085                           (PRE-POST WASH PRODUCTS)
23  0.018827                                     (READY TO EAT)
24  0.061597                    (REFINED EDIBLE OILS-ALL PACKS)
25  0.106071                              (SALTY SNACKS (2/97))
26  0.014009                                 (SANITARY NAPKINS)
27  0.021348                            (SHAMPOO - BY SEGMENTS)
28  0.018976                                      (SKIN CREAMS)
29  0.078867                                   (SPICES (03/04))
30  0.059892                             (TOILET SOAPS (04/00))
31  0.046327                                     (TOOTH PASTES)
32  0.087466                              (VERMICELLI & NOODLE)
33  0.076421                          (WASHING POWDERS/LIQUIDS)
34  0.019717            (BISCUITS - CORE & NON CORE, CHOCOLATE)
35  0.011193   (CLEANERS - UTENSIL, BISCUITS - CORE & NON CORE)
36  0.014676         (BISCUITS - CORE & NON CORE, PACKAGED TEA)
37  0.011341  (REFINED EDIBLE OILS-ALL PACKS, BISCUITS - COR...
38  0.025647  (SALTY SNACKS (2/97), BISCUITS - CORE & NON CORE)
39  0.010674       (SPICES (03/04), BISCUITS - CORE & NON CORE)
40  0.013713  (TOILET SOAPS (04/00), BISCUITS - CORE & NON C...
41  0.012675         (TOOTH PASTES, BISCUITS - CORE & NON CORE)
42  0.024387  (BISCUITS - CORE & NON CORE, VERMICELLI & NOODLE)
43  0.014973  (BISCUITS - CORE & NON CORE, WASHING POWDERS/L...
44  0.014306      (CLEANERS - UTENSIL, WASHING POWDERS/LIQUIDS)
45  0.012156               (TOILET SOAPS (04/00), TOOTH PASTES)
46  0.015937    (TOILET SOAPS (04/00), WASHING POWDERS/LIQUIDS)
47  0.011119            (TOOTH PASTES, WASHING POWDERS/LIQUIDS)

# g.

In [27]:
rules = association_rules(frequent_itemsets2, metric="lift")

In [28]:
rules

antecedents                   consequents  \
0           (CLEANERS - UTENSIL)  (BISCUITS - CORE & NON CORE)   
1   (BISCUITS - CORE & NON CORE)          (CLEANERS - UTENSIL)   
2          (SALTY SNACKS (2/97))  (BISCUITS - CORE & NON CORE)   
3   (BISCUITS - CORE & NON CORE)         (SALTY SNACKS (2/97))   
4         (TOILET SOAPS (04/00))  (BISCUITS - CORE & NON CORE)   
5   (BISCUITS - CORE & NON CORE)        (TOILET SOAPS (04/00))   
6                 (TOOTH PASTES)  (BISCUITS - CORE & NON CORE)   
7   (BISCUITS - CORE & NON CORE)                (TOOTH PASTES)   
8   (BISCUITS - CORE & NON CORE)         (VERMICELLI & NOODLE)   
9          (VERMICELLI & NOODLE)  (BISCUITS - CORE & NON CORE)   
10          (CLEANERS - UTENSIL)     (WASHING POWDERS/LIQUIDS)   
11     (WASHING POWDERS/LIQUIDS)          (CLEANERS - UTENSIL)   
12        (TOILET SOAPS (04/00))                (TOOTH PASTES)   
13                (TOOTH PASTES)        (TOILET SOAPS (04/00))   
14        (TOILET SOAPS (04/00))     (WASHING POWDERS/LIQUIDS)   
15     (WASHING POWDERS/LIQUIDS)        (TOILET SOAPS (04/00))   
16                (TOOTH PASTES)     (WASHING POWDERS/LIQUIDS)   
17     (WASHING POWDERS/LIQUIDS)                (TOOTH PASTES)   

    antecedent support  consequent support   support  confidence      lift  \
0             0.046624            0.283596  0.011193    0.240064  0.846497   
1             0.283596            0.046624  0.011193    0.039467  0.846497   
2             0.106071            0.283596  0.025647    0.241789  0.852581   
3             0.283596            0.106071  0.025647    0.090434  0.852581   
4             0.059892            0.283596  0.013713    0.228960  0.807346   
5             0.283596            0.059892  0.013713    0.048353  0.807346   
6             0.046327            0.283596  0.012675    0.273600  0.964751   
7             0.283596            0.046327  0.012675    0.044694  0.964751   
8             0.283596            0.087466  0.024387    0.085991  0.983135   
9             0.087466            0.283596  0.024387    0.278814  0.983135   
10            0.046624            0.076421  0.014306    0.306836  4.015061   
11            0.076421            0.046624  0.014306    0.187197  4.015061   
12            0.059892            0.046327  0.012156    0.202970  4.381236   
13            0.046327            0.059892  0.012156    0.262400  4.381236   
14            0.059892            0.076421  0.015937    0.266089  3.481870   
15            0.076421            0.059892  0.015937    0.208535  3.481870   
16            0.046327            0.076421  0.011119    0.240000  3.140485   
17            0.076421            0.046327  0.011119    0.145490  3.140485   

    leverage  conviction  
0  -0.002030    0.942715  
1  -0.002030    0.992549  
2  -0.004435    0.944860  
3  -0.004435    0.982808  
4  -0.003272    0.929140  
5  -0.003272    0.987875  
6  -0.000463    0.986238  
7  -0.000463    0.998291  
8  -0.000418    0.998386  
9  -0.000418    0.993368  
10  0.010743    1.332411  
11  0.010743    1.172949  
12  0.009382    1.196534  
13  0.009382    1.274550  
14  0.011360    1.258434  
15  0.011360    1.187808  
16  0.007578    1.215235  
17  0.007578    1.116046

# h.

<div style="direction:rtl">

conviction بالا به این معنی است که نتیجه به شدت به antecedent بستگی دارد. به عنوان مثال، در مورد یک نمره اطمینان کامل، مخرج 0 می شود (به دلیل 1 - 1) که برای آن نمره conviction به عنوان 'inf' تعریف می شود. مشابه lift، اگر موارد independent باشند، conviction 1 است.
    
</div>

<div style="direction:rtl">

Leverage تفاوت بین فرکانس مشاهده‌شده A و C که با هم ظاهر می‌شوند و فرکانسی که اگر A و C independent بودند را محاسبه می‌کند. مقدار اهرم 0 نشان دهنده استقلال است.
    
</div>

<div style="direction:rtl">

مثلا قاعده آخر مشخص میکند که پس از آمدن washing powders/liquids کالای tooth pastes خریداری میشود.
    
ساپورت قاعده ی پیشین برابر با 0.07 و ساپورت قاعده پسین برابر با 0.04 است
    
ساپورت و کانفیدنس قاعده برابر با 0.01 و 0.14 است.
    
مقدار lift برابر با 3.14 مقدار liverage برابر با 0.007 و مقدار conviction برابر با 1.11 میباشد.
    
</div>

# i.

In [29]:
rules[ (rules['confidence'] > 0.3) &
       (rules['lift'] > 2) ]

antecedents                consequents  antecedent support  \
10  (CLEANERS - UTENSIL)  (WASHING POWDERS/LIQUIDS)            0.046624   

    consequent support   support  confidence      lift  leverage  conviction  
10            0.076421  0.014306    0.306836  4.015061  0.010743    1.332411